### ISSUE TITLE
Use webpack-merge in place of webpacker's custom config helpers

### ISSUE THREAD

**Issue by 0xdevalias** [rails/webpacker#2376]
Due to the way the current 'base' webpack settings are serialised in `toWebpackConfig()` it can be kind of annoying to try and work with in the 'webpacker config' way. You basically have to intricately understand which helper methods are used, and what paths they map to in:

- https://github.com/rails/webpacker/blob/master/package/environments/base.js#L113-L119

For example, I would have to change resolve plugins I would use:
- `environment.config.resolve.plugins`
- `environment.config.resolveLoader.plugins`

But there are other plugins in:
- `environment.plugins`

Etc.

----

It feels like it would be a lot simpler to just work directly with the webpack config layout 'as it is', and I think webpack-merge makes that a lot easier:

- https://github.com/survivejs/webpack-merge

It's already suggested on the webpack config docs page:

- https://github.com/rails/webpacker/blob/master/docs/webpack.md#configuration

And it seems to be recommended here: https://webpack.js.org/guides/production/

> While we will separate the production and development specific bits out, note that we'll still maintain a "common" configuration to keep things DRY. In order to merge these configurations together, we'll use a utility called webpack-merge. With the "common" configuration in place, we won't have to duplicate code within the environment-specific configurations.

I realise to replace the current config setup entirely would be a breaking change, but perhaps an adapter could allow people to progressively 'opt in' in the meantime.

----

My current workaround/manual usage:

`config/webpack/environment.js`:
```js
// Load the base @rails/webpacker configuration object + base environment overrides
const { environment } = require('@rails/webpacker')

const webpackMerge = require('webpack-merge')
const path = require('path')
const webpack = require('webpack')

// -------------------------------------------------------------------------------
//                                WARNING
// -------------------------------------------------------------------------------
//   Do any @rails/webpacker specific config/helper/weirdness changes before
//   here.. from this point on we're just using normal webpack config JS objects
// -------------------------------------------------------------------------------
const webpackerDefaultWebpackConfig = environment.toWebpackConfig().toObject()

const mergeData = {
  // https://webpack.js.org/configuration/resolve/
    modules: [
      path.resolve(__dirname, '../../node_modules'),

      // TODO: For now we read this from @rails/webpacker's yaml config to avoid confusion
      //  but when we move away from @rails/webpacker replace it with the commented line
      // path.resolve(__dirname, '../../app/javascript'),
      path.resolve(__dirname, '../../', webpackerYamlConfig.source_path),
    ],
  },
}

const mergeStrategies = {
  'resolve.modules': 'replace',
}

// Ref: https://github.com/survivejs/webpack-merge#merging-with-strategies
const mergedWebpackConfig = webpackMerge.smartStrategy(mergeStrategies)(
  webpackerDefaultWebpackConfig,
  mergeData
)

module.exports = mergedWebpackConfig
```

Then for environment specific overrides, we can similarly work with it like so:

`config/webpack/development.js`
```js
const webpackMerge = require('webpack-merge')
const baseWebpackConfig = require('./environment')

// NOTE: The config in this file is plain serialised webpack JS config, not @rails/webpacker magic
// Ref: https://webpack.js.org/configuration/
const mergeData = {
  // Ref: https://webpack.js.org/configuration/dev-server/
  devServer: {
    stats: 'errors-only',
  },
}

module.exports = webpackMerge(baseWebpackConfig, mergeData)
```

This would allow the webpacker/webpack config files to be simpler/more 'canonicly webpack', which would allow other tools that consume/parse webpack config to work in a more natural way.

It would also IMO make it easier to use newer webpack features like:

- https://webpack.js.org/configuration/configuration-types/#exporting-a-function

As it would align webpacker to using webpack config/exports in a more 'normal' way.

**Comment by justin808** [comment]
@jakeNiemiec @gauravtiwari I think this is a great idea.

**Comment by jakeNiemiec** [comment]
I agree. I have always envisioned a way to statically get and set the webpack config for devs who know what they are doing, and just want tell webpack(er) exactly what config to use without losing all of the other helpful stuff that webpack provides.

**Comment by 0xdevalias** [comment]
I'm curious if you have any particular patterns in mind for how best to support/implement this? And whether maintaining backwards compatibility with current methods would be desired in doing so?

**Comment by sergiopantoja** [comment]
FYI seems like `webpack-merge` is being officially adopted. From the README:

```
// config/webpack/base.js
const { webpackConfig, merge } = require('@rails/webpacker')
const customConfig = require('./custom')

module.exports = merge(webpackConfig, customConfig)
```
